library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,269 @@
+//! code conforms to strict provenance.
+//!
+//!
+//! ## Using Strict Provenance
+//!
+//! Most code needs no changes to conform to strict provenance, as the only really concerning
+//! operation that *wasn't* obviously already Undefined Behaviour is casts from usize to a
+//! pointer. For code which *does* cast a usize to a pointer, the scope of the change depends
+//! on exactly what you're doing.
+//!
+//! In general you just need to make sure that if you want to convert a usize address to a
+//! pointer and then use that pointer to read/write memory, you need to keep around a pointer
+//! that has sufficient provenance to perform that read/write itself. In this way all of your
+//! casts from an address to a pointer are essentially just applying offsets/indexing.
+//!
+//! This is generally trivial to do for simple cases like tagged pointers *as long as you
+//! represent the tagged pointer as an actual pointer and not a usize*. For instance:
+//!
+//! ```
+//! #![feature(strict_provenance)]
+//!
+//! unsafe {
+//!     // A flag we want to pack into our pointer
+//!     static HAS_DATA: usize = 0x1;
+//!     static FLAG_MASK: usize = !HAS_DATA;
+//!
+//!     // Our value, which must have enough alignment to have spare least-significant-bits.
+//!     let my_precious_data: u32 = 17;
+//!     assert!(core::mem::align_of::<u32>() > 1);
+//!
+//!     // Create a tagged pointer
+//!     let ptr = &my_precious_data as *const u32;
+//!     let tagged = ptr.map_addr(|addr| addr | HAS_DATA);
+//!
+//!     // Check the flag:
+//!     if tagged.addr() & HAS_DATA != 0 {
+//!         // Untag and read the pointer
+//!         let data = *tagged.map_addr(|addr| addr & FLAG_MASK);
+//!         assert_eq!(data, 17);
+//!     } else {
+//!         unreachable!()
+//!     }
+//! }
+//! ```
+//!
+//! (Yes, if you've been using AtomicUsize for pointers in concurrent datastructures, you should
+//! be using AtomicPtr instead. If that messes up the way you atomically manipulate pointers,
+//! we would like to know why, and what needs to be done to fix it.)
+//!
+//! Something more complicated and just generally *evil* like a XOR-List requires more significant
+//! changes like allocating all nodes in a pre-allocated Vec or Arena and using a pointer
+//! to the whole allocation to reconstitute the XORed addresses.
+//!
+//! Situations where a valid pointer *must* be created from just an address, such as baremetal code
+//! accessing a memory-mapped interface at a fixed address, are an open question on how to support.
+//! These situations *will* still be allowed, but we might require some kind of "I know what I'm
+//! doing" annotation to explain the situation to the compiler. It's also possible they need no
+//! special attention at all, because they're generally accessing memory outside the scope of
+//! "the abstract machine", or already using "I know what I'm doing" annotations like "volatile".
```

Comment:
In a bare-metal CHERI world, programs start out with an all-powerful capability (e.g., at hardware reset the default data capability (DDC) and program counter capability (PCC) are initialized to such values). It's up to the language runtime's startup code to subset that into appropriate capabilities for users of C/Rust/etc. For example in CheriBSD, we chop the root capability into a number of sub-capabilities in `cheri_init_capabilities` which is called from `_start` (https://github.com/CTSRD-CHERI/cheribsd/blob/2681a143af26941d54997cfd47e645dcdfbdf3ea/sys/arm64/cheri/cheri_machdep.c#L49). I'd expect a Rust environment targeting bare metal to expose similar capabilities to create pointers for MMIO access and the like. How to expose those capabilities will surely require some exploration.

P.S. The split linked above is less complete that ideal (e.g., the kernel root overlaps with user space), but the split allows us to express our intent and should eventually let us tighten things down further.

Reply:
Yeah I have been actively resisting blessing an "integers cast to pointers get a fresh valid provenance that is sound as long as it doesn't conflict with memory 'inside' the abstract machine" interpretation of the MMIO case, precisely because it seems hostile to CHERI which wants a bit more information/explicitness.